modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaDescriptor.java

```diff
@@ -84,71 +82,118 @@ public class SchemaDescriptor {
-                .forEach(c -> {
-                    if (c.type() instanceof TemporalNativeType) {
-                        hasTemporalColumns.set(true);
-                    }
-
-                    colMap.put(c.name(), c);
-                });
-
-        this.physicalOrderMatchesLogical = colMap.values().stream().allMatch(col -> col.columnOrder() == col.schemaIndex());
-
-        this.hasTemporalColumns = hasTemporalColumns.get();
-
-        // Preserving key chunk column order is not actually required.
-        // It is sufficient to has same column order for all nodes.
-        if (ArrayUtils.nullOrEmpty(colocationCols)) {
-            this.colocationCols = this.keyCols.columns();
-            this.colocationColIndexes = null;
-        } else {
-            this.colocationCols = new Column[colocationCols.length];
-            this.colocationColIndexes = new HashMap<>(colocationCols.length);
-
-            for (int i = 0; i < colocationCols.length; i++) {
-                Column col = colMap.get(colocationCols[i]);
-                this.colocationCols[i] = col;
-                this.colocationColIndexes.put(col, i);
+        if (colocationColumns == null) {
+            colocationColumns = keyColumns;
+        }
+
+        Object2IntMap<String> columnNameToPositionInColocation = new Object2IntOpenHashMap<>();
+        idx = 0;
+        for (String name : colocationColumns) {
+            assert columnNameToPositionInKey.containsKey(name)
+                    : "Colocation column must be part of the key: " + name;
+            assert !columnNameToPositionInColocation.containsKey(name)
+                    : "Colocation column must not have duplicates: " + name;
+
+            columnNameToPositionInColocation.put(name, idx++);
+        }
+
+        boolean hasTemporalColumns = false;
+        int rowPosition = 0;
+        int valuePosition = 0;
+        for (Column column : columns) {
+            Column orderedColumn = column.copy(
+                    rowPosition++,
+                    columnNameToPositionInKey.getOrDefault(column.name(), -1),
+                    columnNameToPositionInKey.containsKey(column.name()) ? -1 : valuePosition++,
+                    columnNameToPositionInColocation.getOrDefault(column.name(), -1)
+            );
+
+            Column old = columnsByName.put(orderedColumn.name(), orderedColumn);
+
+            assert old == null : "Columns with similar names are not allowed: " + old.name();
+
+            orderedColumns.add(orderedColumn);
+
+            if (column.type() instanceof TemporalNativeType) {
+                hasTemporalColumns = true;
             }
         }
 
-        // TODO: Move keyIndex and colocationIndex to Column class for faster and simpler access?
-        keyColIndexes = new HashMap<>(keyCols.length);
+        this.ver = ver;
+        this.columns = List.copyOf(orderedColumns);
+        this.columnsByName = Map.copyOf(columnsByName);
```

Comment:
Can a copy be replaced with immutable collection?

Reply:
according to javadoc, `Map.copyOf` returns unmodifiable Map